src/libraries/System.Utf8String.Experimental/src/System/Utf8String.Portable.cs

```diff
@@ -0,0 +1,231 @@
+        /// <returns></returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal ref byte DangerousGetMutableReference() =>
+            ref MemoryMarshal.GetReference(_bytes.AsSpan());
+
+        /// <summary>
+        /// Returns a <em>mutable</em> <see cref="Span{Byte}"/> that can be used to populate this
+        /// <see cref="Utf8String"/> instance. Only to be used during construction.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal Span<byte> DangerousGetMutableSpan()
+        {
+            Debug.Assert(Length > 0, $"This should only ever be called on a non-empty {nameof(Utf8String)}.");
+            return _bytes.AsSpan(0, Length);
+        }
+
+        /// <summary>
+        /// Returns a <see cref="ReadOnlySpan{Byte}"/> for this
+        /// <see cref="Utf8String"/> instance.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal ReadOnlySpan<byte> GetSpan() => _bytes.AsSpan(0, Length);
+
+        /// <summary>
+        /// Gets an immutable reference that can be used in a <see langword="fixed"/> statement. The resulting
+        /// reference can be pinned and used as a null-terminated <em>LPCUTF8STR</em>.
+        /// </summary>
+        /// <remarks>
+        /// If this <see cref="Utf8String"/> instance is empty, returns a reference to the null terminator.
+        /// </remarks>
+        [EditorBrowsable(EditorBrowsableState.Never)] // for compiler use only
+        public ref readonly byte GetPinnableReference() => ref _bytes.AsSpan().GetPinnableReference();
+
+        /// <summary>
+        /// Similar to <see cref="Utf8Extensions.AsBytes(Utf8String)"/>, but skips the null check on the input.
+        /// Throws a <see cref="NullReferenceException"/> if the input is null.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal ReadOnlySpan<byte> AsBytesSkipNullCheck() => GetSpan();
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal ReadOnlyMemory<byte> CreateMemoryBytes(int start, int length) =>
+            _bytes.AsMemory(start, length);
+
+        /// <summary>
+        /// Creates a new zero-initialized instance of the specified length. Actual storage allocated is "length + 1" bytes
+        /// because instances are null-terminated.
+        /// </summary>
+        /// <remarks>
+        /// The implementation of this method checks its input argument for overflow.
+        /// </remarks>
+        private static Utf8String FastAllocate(int length)
+        {
+            // just simulate a "fast allocate", since this is portable
+            return new Utf8String(AllocateBuffer(length));
+        }
+
+        private static byte[] AllocateBuffer(int length)
+        {
+            // Actual storage allocated is "length + 1" bytes because instances are null-terminated.
+            return new byte[length + 1];
```

Comment:
Check for negative _length_ inputs. Otherwise somebody could pass a negative value which given the +1 could result in a zero-length array being allocated.

Reply:
I'll also add a check for `int.MaxValue`, so this doesn't overflow.

Reply:
I've analyzed all the call sites. The `length` will always be positive, so I am going to add that as an assert since it should never happen.
There is a case where `int.MaxValue` could come into this method, so I'll add a check for that, and subtract 1 in that case, since the VM will just OOM if you try to create a `byte[int.MaxValue]`.